### ISSUE TITLE
Planning: Higher level API endpoints and CRUD functions

### ISSUE THREAD

**Issue by mtreacy002** [anitab-org/bridge-in-tech-backend#15]
[As per proposal](https://docs.google.com/document/d/1ZZCOoWdn2yb6N3qrrgK0SRDQPcx2Pp7kz3yDXDRSHUs/edit#heading=h.8ii92113maf5):

**Higher Level API (CRUD operations)**

![Screen Shot 2020-05-11 at 3.06.53 pm.png](https://images.zenhubusercontent.com/5eb4e4b9636d69f8661b2c28/1a008d5e-be90-477a-87a2-593ea76e1d37)
![Screen Shot 2020-05-11 at 3.07.30 pm.png](https://images.zenhubusercontent.com/5eb4e4b9636d69f8661b2c28/fc7c5237-a947-48f0-a606-2e6eedd3e5fd)
![Screen Shot 2020-05-11 at 3.08.12 pm.png](https://images.zenhubusercontent.com/5eb4e4b9636d69f8661b2c28/8b4eeb68-4f90-42ad-b268-d90d6dabf27e)
![Screen Shot 2020-05-11 at 3.08.45 pm.png](https://images.zenhubusercontent.com/5eb4e4b9636d69f8661b2c28/50d497a2-a3f0-40a1-a2ec-066817530a38)
![Screen Shot 2020-05-11 at 3.10.46 pm.png](https://images.zenhubusercontent.com/5eb4e4b9636d69f8661b2c28/052722ed-a026-49eb-8c1d-348f9277cf4d)
![Screen Shot 2020-05-11 at 3.11.13 pm.png](https://images.zenhubusercontent.com/5eb4e4b9636d69f8661b2c28/43f58f47-81c3-4ca0-992c-a412a515ba27)

**REST API endpoints examples**

POST /user/personal_background
=> this will add personal background information to the user profile.
GET, PUT, DELETE /user/personal_background/{user_id}
=> this will get/update/delete personal background information of the user which id is specified.
GET /users/find_by_physical_ability/{physical_ability}
=> this will get a list of all users which have the physical ability match the type specified in the ‘physical_ability’ passed as parameter.
GET /users/find_by_criteria/{criteria}
=> this will get a list of users that met the specific search criteria set in ‘criteria’ parameter which is a map of key-value (or dictionary). This way we don’t need to write an extensive list of GETs to filter users by each of their attributes (such as GET /users/find_by_physical_ability).
POST /company
=> this will create a company.
GET, PUT, DELETE /company/{company_id}
=> this will get/update/delete the details of the company which id is specified.
GET /companies
=> this will get the list of all companies.
GET /companies/{company_id}
=> this will get the company which id is specified.
GET /companies/find_by_country/{country}
=> this will get a list of companies in the specified country.
GET /companies/find_by_criteria/{criteria}
=> this will get a list of companies that met the specific search criteria set in ‘criteria’ parameter which is a map of key-value (or dictionary). 
POST /program
=> this will create a program.
GET/PUT/DELETE /program/{program_id}
=> this will get/update/delete the details of the program with the specified id.
GET /programs
=> this will get the list of all programs.
GET /programs/{program_id}
=> this will get the program which id is specified.
GET /programs/find_by_contact_type/{contact_type}
=> this will get a list of programs that have contact_type specified by the parameter.
GET /programs/find_by_criteria/{criteria}
=> this will get a list of programs that met the specific search criteria set in ‘criteria’ parameter which is a map of key-value (~ dictionary).
POST /mentorship_relation/send_request
=> this will create a mentoring relation request as well as send email notification to recipient about MR request
GET, DELETE /mentorship_relation/{mentorship_relation_id}
=> this will get/delete the mentorship_relation with the specified id.
GET /mentorship_relations
=> this will get the list of all MRs.
PUT /mentorship_relations/{mentorship_relation_id}
/accept
=> this will allow the Company/Mentor/Mentee to accept an MR request.
PUT /mentorship_relations/{mentorship_relation_id}
/reject
=> this will allow the Company/Mentor/Mentee to reject an MR request.
PUT /mentorship_relations/{mentorship_relation_id}
/cancel
=> this will allow the Company/Mentor/Mentee to cancel an MR request.
GET / mentorship_relations/current
=> this will get the list of current MR regardless of their ‘state’ for the logged in user.
GET / mentorship_relations/find_by_criteria/{criteria}
=> this will get the list of MRs based on the ‘criteria’ passed on as the parameter.

**Comment by mtreacy002** [comment]
Hi @anitab-org/bridgeintech-maintainers, below is the screenshot of what I have in mind for API endpoints and their related CRUD operations in relation to Users API service. Please give me feedback to whether or not this makes sense to you all.

<img width="550" alt="Screen Shot 2020-05-16 at 7 02 00 am" src="https://user-images.githubusercontent.com/29667122/82095858-41c11f80-9743-11ea-9b00-0fdeae9980a5.png">
<img width="551" alt="Screen Shot 2020-05-16 at 7 02 14 am" src="https://user-images.githubusercontent.com/29667122/82095882-4ab1f100-9743-11ea-9b40-c02a295d735c.png">

[Link to file](https://docs.google.com/document/d/1uCDCWs8Xyo-3EaUnrLOs48mefXJIN235b1aAswA-s-Y/edit#heading=h.kuvjowaa2l13)

Note: I haven't decided on the API endpoints for `Home` and `Dashboard` yet because it'll come after I get more picture on the whole API endpoints

**Comment by ramitsawhney27** [comment]
For the initial few endpoints, as you start implementing them, you should be documenting the error codes. For example: 400 bad request - and when would you return this?

**Comment by mtreacy002** [comment]
Update:
Hi @anitab-org/bridgeintech-maintainers. Here's the updated API endpoints and error handlings related to Users service

<img width="564" alt="Screen Shot 2020-05-29 at 10 36 25 am" src="https://user-images.githubusercontent.com/29667122/83208214-6a600500-a198-11ea-92f1-2ed1fa8a1040.png">
<img width="564" alt="Screen Shot 2020-05-29 at 10 38 02 am" src="https://user-images.githubusercontent.com/29667122/83208260-81065c00-a198-11ea-8964-f186997e1e9f.png">
<img width="562" alt="Screen Shot 2020-05-29 at 10 38 55 am" src="https://user-images.githubusercontent.com/29667122/83208291-9ed3c100-a198-11ea-90b8-fcd286bcbbc2.png">
<img width="564" alt="Screen Shot 2020-05-29 at 10 39 20 am" src="https://user-images.githubusercontent.com/29667122/83208314-ae530a00-a198-11ea-8d08-108150105b03.png">
<img width="565" alt="Screen Shot 2020-05-29 at 10 39 49 am" src="https://user-images.githubusercontent.com/29667122/83208338-c1fe7080-a198-11ea-9d03-aab799f5ff82.png">

**Comment by ramitsawhney27** [comment]
These look good. What if there is an internal server error?

**Comment by mtreacy002** [comment]
I'm thinking to mask it with response 500 but display the error message as it would catch by the HTTPError exception. As mentioned in the below

<img width="991" alt="Screen Shot 2020-05-29 at 8 01 14 pm" src="https://user-images.githubusercontent.com/29667122/83247666-36f89700-a1e7-11ea-8bc6-c6809bcdb120.png">

Would this be Ok, @ramitsawhney27? If you agree to this, I would add error code 500 on it as well with unspecified message (coz it'll be whatever returned by the response)

**Comment by mtreacy002** [comment]
Update:
I've added error response 500 in all API requests because of the possibilities of internal error can happen anywhere in the request.
Below just an example of what I've added.

<img width="563" alt="Screen Shot 2020-05-29 at 8 24 43 pm" src="https://user-images.githubusercontent.com/29667122/83249786-74125880-a1ea-11ea-9c6b-7631c07efb83.png">

**Comment by meenakshi-dhanani** [comment]
Hey Maya, 
Everything looks good. Just a question to think about the login api. So if there's any issue with authentication it makes sense to return a 401. Find explanation in the RFC [here](https://tools.ietf.org/html/rfc7235#section-3.1). 

And in case the credentials a wrong, if you send back a 404, it might give a tip to the malicious users, to try combinations. A 401 should work for any invalid credentials.

**Comment by mtreacy002** [comment]
Hi @meenakshi-dhanani. That also makes sense to me, tbh (although I'm not sure about the 404 giving more tip to malicious user more than 401 does, 😃). However, in terms of error code, as BIT is using MS API service for this one (login), I don't know if it's a good idea to change the error code from 404 (given by the MS API) to 401 on BIT UI to client. Shouldn't BIT only passing whatever value is returned by MS API? 
@ramitsawhney27 and @isabelcosta, is there any reason to use 404 instead of 401 on MS API on this matter? Is BIT allowed to change this?

**Comment by meenakshi-dhanani** [comment]
Couple of points:

401 can take care of any invalid credentials (credentials do not exist, or
they are invalid). So 401 can be a blanket status code for that.

If there is another response, say 404. Assuming I am a malicious user, I
attempt to login with some credentials, if I get a 404 I know that these do
not exist, so I'll try with another pair of credentials and might get a 401
and then I can continue and try till I get the right password for it.

Since, 401 can take care of invalid I say we return only a 401 for any
issue with credentials. Does that make sense?

**Comment by mtreacy002** [comment]
Update:
Here's the API endpoints and errors handling for Organizations service

<img width="565" alt="Screen Shot 2020-05-30 at 4 15 53 pm" src="https://user-images.githubusercontent.com/29667122/83321178-21d84280-a291-11ea-9e13-f65b0a1791ba.png">
<img width="563" alt="Screen Shot 2020-05-30 at 4 16 12 pm" src="https://user-images.githubusercontent.com/29667122/83321187-2dc40480-a291-11ea-95d2-c40ec6df7cca.png">
<img width="564" alt="Screen Shot 2020-05-30 at 4 16 43 pm" src="https://user-images.githubusercontent.com/29667122/83321220-65cb4780-a291-11ea-890d-e653ce53bd1a.png">
<img width="564" alt="Screen Shot 2020-05-30 at 4 22 14 pm" src="https://user-images.githubusercontent.com/29667122/83321264-c5295780-a291-11ea-9d0f-06fea67d37ba.png">

**Comment by mtreacy002** [comment]
Update:

So, Here're the latest version for API endpoints

User service

<img width="564" alt="Screen Shot 2020-05-31 at 12 43 26 am" src="https://user-images.githubusercontent.com/29667122/83331184-e1e87e00-a2d7-11ea-81f5-1d8f4c7b7959.png">
<img width="564" alt="Screen Shot 2020-05-31 at 12 44 42 am" src="https://user-images.githubusercontent.com/29667122/83331197-f2005d80-a2d7-11ea-9966-2b198a95f98a.png">
<img width="566" alt="Screen Shot 2020-05-31 at 12 45 07 am" src="https://user-images.githubusercontent.com/29667122/83331207-017fa680-a2d8-11ea-8d33-ea146636f320.png">
<img width="563" alt="Screen Shot 2020-05-31 at 12 45 31 am" src="https://user-images.githubusercontent.com/29667122/83331217-0fcdc280-a2d8-11ea-8eb1-c23a4dbe5d82.png">
<img width="562" alt="Screen Shot 2020-05-31 at 12 46 01 am" src="https://user-images.githubusercontent.com/29667122/83331227-2116cf00-a2d8-11ea-9045-1193ab9e01ac.png">

![Screen Shot 2020-05-31 at 12 47 26 am](https://user-images.githubusercontent.com/29667122/83331262-591e1200-a2d8-11ea-95a1-247769fcbf2b.png)

![Screen Shot 2020-05-31 at 12 47 41 am](https://user-images.githubusercontent.com/29667122/83331265-5d4a2f80-a2d8-11ea-856e-4d186bfd6d4a.png)

**Comment by mtreacy002** [comment]
Update:

Organizations service

![Screen Shot 2020-05-31 at 12 48 08 am](https://user-images.githubusercontent.com/29667122/83331294-9d111700-a2d8-11ea-97e8-725b59d7f357.png)
![Screen Shot 2020-05-31 at 12 50 00 am](https://user-images.githubusercontent.com/29667122/83331308-b7e38b80-a2d8-11ea-830f-4447c770aac0.png)
![Screen Shot 2020-05-31 at 12 50 55 am](https://user-images.githubusercontent.com/29667122/83331328-d47fc380-a2d8-11ea-952b-9ca3f4c5904d.png)
![Screen Shot 2020-05-31 at 12 51 29 am](https://user-images.githubusercontent.com/29667122/83331332-e3667600-a2d8-11ea-97b6-33ed260c3703.png)

**Comment by mtreacy002** [comment]
Update:

Programs service

![Screen Shot 2020-05-31 at 12 51 52 am](https://user-images.githubusercontent.com/29667122/83331341-fe38ea80-a2d8-11ea-9113-d22444c844aa.png)
![Screen Shot 2020-05-31 at 12 52 45 am](https://user-images.githubusercontent.com/29667122/83331350-10b32400-a2d9-11ea-950c-d3eb4796bf6b.png)
![Screen Shot 2020-05-31 at 12 53 09 am](https://user-images.githubusercontent.com/29667122/83331356-245e8a80-a2d9-11ea-9eab-95aec0ca5005.png)

**Comment by mtreacy002** [comment]
Update:
API endpoints for Mentorship_Relations service

<img width="566" alt="Screen Shot 2020-05-31 at 8 47 43 pm" src="https://user-images.githubusercontent.com/29667122/83350542-1fa0e180-a380-11ea-964f-da3c699217a8.png">
<img width="568" alt="Screen Shot 2020-05-31 at 8 49 04 pm" src="https://user-images.githubusercontent.com/29667122/83350555-2f202a80-a380-11ea-9cbf-583721fcc20d.png">
<img width="562" alt="Screen Shot 2020-05-31 at 8 49 25 pm" src="https://user-images.githubusercontent.com/29667122/83350566-3e9f7380-a380-11ea-9ffb-27d128f80f94.png">
<img width="565" alt="Screen Shot 2020-05-31 at 8 50 06 pm" src="https://user-images.githubusercontent.com/29667122/83350577-57a82480-a380-11ea-8f95-864665a2ee56.png">
<img width="564" alt="Screen Shot 2020-05-31 at 8 50 37 pm" src="https://user-images.githubusercontent.com/29667122/83350587-6d1d4e80-a380-11ea-8bb7-43cdd837844a.png">
<img width="562" alt="Screen Shot 2020-05-31 at 8 51 10 pm" src="https://user-images.githubusercontent.com/29667122/83350598-7efef180-a380-11ea-8884-085df2711193.png">
<img width="564" alt="Screen Shot 2020-05-31 at 8 51 39 pm" src="https://user-images.githubusercontent.com/29667122/83350606-90e09480-a380-11ea-92b9-03f54d492c6c.png">
<img width="562" alt="Screen Shot 2020-05-31 at 8 52 12 pm" src="https://user-images.githubusercontent.com/29667122/83350615-9f2eb080-a380-11ea-9cd6-aa7b00761d51.png">
<img width="563" alt="Screen Shot 2020-05-31 at 8 52 37 pm" src="https://user-images.githubusercontent.com/29667122/83350621-b2418080-a380-11ea-8487-a7c97572cb03.png">
<img width="565" alt="Screen Shot 2020-05-31 at 8 53 08 pm" src="https://user-images.githubusercontent.com/29667122/83350631-c2596000-a380-11ea-89d2-9a5b41b5a3d8.png">
<img width="563" alt="Screen Shot 2020-05-31 at 8 53 31 pm" src="https://user-images.githubusercontent.com/29667122/83350639-cf764f00-a380-11ea-9d31-68ab44c2f1bc.png">
<img width="565" alt="Screen Shot 2020-05-31 at 8 53 54 pm" src="https://user-images.githubusercontent.com/29667122/83350643-dc933e00-a380-11ea-9732-1831423acfc8.png">
<img width="562" alt="Screen Shot 2020-05-31 at 8 54 16 pm" src="https://user-images.githubusercontent.com/29667122/83350655-eae15a00-a380-11ea-8d8b-f4600094cfb6.png">
<img width="562" alt="Screen Shot 2020-05-31 at 8 54 42 pm" src="https://user-images.githubusercontent.com/29667122/83350661-f92f7600-a380-11ea-84cf-fadc32406cc9.png">

**Comment by ramitsawhney27** [comment]
Hi @mtreacy002, your error codes and specs look great. Thank you for the great work with them. One small suggestion, could we use "201-Created", rather than "200-OK" as the error code if something new is created? This is as per our previous discussions in weekly meetings.

**Comment by mtreacy002** [comment]
Sorry, haven't updated the docs yet, but have actually applied your suggestion in the code base for User Registration ([PR#26](https://github.com/anitab-org/bridge-in-tech-backend/pull/26)). 
![Screen Shot 2020-06-03 at 9 15 17 am](https://user-images.githubusercontent.com/29667122/83578776-f1d3bc80-a57a-11ea-8632-31ad46317be5.png)

Will change the docs now... 😉

**Comment by mtreacy002** [comment]
Hi @anitab-org/bridgeintech-maintainers. I have updated the docs for the REST API Endpoints and errors handling [here](https://docs.google.com/document/d/1uCDCWs8Xyo-3EaUnrLOs48mefXJIN235b1aAswA-s-Y/edit#heading=h.kuvjowaa2l13). Also, I highlighted 18 error codes that I think would require changes to be done on Mentorship System backend side. This is because if we change it on MS backend, it only needs to change the word selection on HTTPStatus itself whereas if we want to keep MS as it is and mask the codes on BIT side, we would need to write extra lines of codes (e.g. `if` condition and even sometimes doing `for` loop) for each of the error points. 
Please let me know what you think.

**Comment by meenakshi-dhanani** [comment]
Ideally, if MS has the logic of the API, then MS error codes need to be appropriate. Let's open issues on MS, if we can pick them up because this affects BIT too, let's do that. If we don't fix MS, there will be some unnecessary code on BIT, which might have to be removed once MS is fixed, so let's fix the source(because we can). Create them as issues on MS I say, if we can let's help someone who picks it up or let's pick it up. (Don't be blocked by it though)

**Comment by mtreacy002** [comment]
I agree with @meenakshi-dhanani .
@ramitsawhney27, @isabelcosta and @foongminwong. would you be ok if I open the issues on MS backend to fix the misrepresented http error codes?
Note that the 18 highlighted error codes that need to be replaced must be represented in separate issue (individually), or at least per function (e.g. user register function) so not to block the BIT development coz the changes can be done gradually.

**Comment by mtreacy002** [comment]
Update @anitab-org/bridgeintech-maintainers . Closing this as discussion on API endpoints will be ongoing on [BridgeInTech Proposal Review doc](https://docs.google.com/document/d/1uCDCWs8Xyo-3EaUnrLOs48mefXJIN235b1aAswA-s-Y/edit#heading=h.kuvjowaa2l13)
